Implement feature-flagged support for ignoring of optional array fields with `minItems` set

RPiAwesomeness opened the pull request:
### Reasons for making this change

Currently, if `minItems` is set on an optional field that field will show up as an empty array in the form data if no values have been entered by the user. This is an area of ambiguity in the spec and some users want this behavior while others don't. Thus, it should be possible to choose whether you want the field to be populated as an empty array or to be ignored and thus have the key non-existent in the form output if no items have been added.

- fixes #3363
- fixes unreported bug where live settings in the Playground aren't being saved/loaded

### Checklist

- [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests. I've run `npm run test:update` to update snapshots, if needed.
  - [x] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [x] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR

RPiAwesomeness commented:
Sorry about the commit mess. Wanted to make sure my commits were being verified properly.

heath-freenome commented on CHANGELOG.md:
```diff
@@ -16,6 +16,20 @@ should change the heading of the (upcoming) version to include a major version b
 
 -->
 
+# 5.6.1
```
You are adding a new feature so let's bump the version
```suggestion
# 5.7.0
```

RPiAwesomeness replied:
Done

heath-freenome commented on CHANGELOG.md:
```diff
@@ -16,6 +16,20 @@ should change the heading of the (upcoming) version to include a major version b
 
+# 5.6.1
+
+## @rjsf/core & @rjsf/utils
```
You already mention `@rjsf/util` below
```suggestion
## @rjsf/core
```

RPiAwesomeness replied:
Ah, that's leftover from a previous set of CHANGELOG changes, my bad

heath-freenome commented on packages/core/src/components/Form.tsx:
```diff
@@ -305,9 +309,12 @@ export default class Form<
     const mustValidate = edit && !props.noValidate && liveValidate;
     const rootSchema = schema;
+    const behaviorBitFlags =
+      ('behaviorBitFlags' in props ? props.defaultFormStateBehavior : this.props.defaultFormStateBehavior) ||
```
I think you missed this when you renamed the prop
```suggestion
      ('defaultFormStateBehavior' in props ? props.defaultFormStateBehavior : this.props.defaultFormStateBehavior) ||
```

heath-freenome reviewed (commented):
Sorry I've been busy. This is a partial review. Will dig deeper tomorrow

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -27,26 +27,30 @@ import {
 {
   rootSchema: S;
   validator: ValidatorType<T, S, F>;
+  behaviorBitFlags: number;
```
```suggestion
  defaultFormStateBehavior: number;
```

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -27,26 +27,30 @@ import {
    *
    * @param validator - An implementation of the `ValidatorType` interface that will be forwarded to all the APIs
    * @param rootSchema - The root schema that will be forwarded to all the APIs
+   * @param behaviorBitFlags Bitwise flags to set which behavior is chosen for certain edge cases.
```
```suggestion
   * @param [defaultFormStateBehavior] - Optional bit flag, if provided, allows users to override default form state behavior
```

RPiAwesomeness replied:
I'm open to discuss this but I had intentionally kept this parameter as required for the constructor here. Making the parameter for the `SchemaUtils` constructor makes it clear that this is important to the usage of `SchemaUtils`. Also, when I made it optional, it caused the unit tests to become more complex without providing much, if any, benefit. If a user wants to create a `SchemaUtils` instance I think it makes sense for them to explicitly specify which behavior they want. 

The only place this object is constructed is `createSchemaUtils()` in `createSchemaUtils.ts:289` and the parameter for that function is optional, which I believe accomplishes the same result from the perspective of the library's users.

heath-freenome replied:
Oh, yeah I agree that making it required is necessary, I was just copy/pasting the name and documentation from other lines and missed this one.

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -27,26 +27,30 @@ import {
+   * @param behaviorBitFlags Bitwise flags to set which behavior is chosen for certain edge cases.
    */
-  constructor(validator: ValidatorType<T, S, F>, rootSchema: S) {
+  constructor(validator: ValidatorType<T, S, F>, rootSchema: S, behaviorBitFlags: number) {
     this.rootSchema = rootSchema;
     this.validator = validator;
+    this.behaviorBitFlags = behaviorBitFlags;
```
```suggestion
  constructor(validator: ValidatorType<T, S, F>, rootSchema: S, defaultFormStateBehavior: number) {
    this.rootSchema = rootSchema;
    this.validator = validator;
    this. defaultFormStateBehavior = defaultFormStateBehavior;
```

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -63,13 +67,18 @@ class SchemaUtils<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends Fo
    *
    * @param validator - An implementation of the `ValidatorType` interface that will be compared against the current one
    * @param rootSchema - The root schema that will be compared against the current one
+   * @param behaviorBitFlags Bitwise flags to set which behavior is chosen for certain edge cases
```
```suggestion
   * @param [defaultFormStateBehavior] - Optional bit flag, if provided, allows users to override default form state behavior
```

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -63,13 +67,18 @@ class SchemaUtils<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends Fo
    * @returns - True if the `SchemaUtilsType` differs from the given `validator` or `rootSchema`
    */
-  doesSchemaUtilsDiffer(validator: ValidatorType<T, S, F>, rootSchema: S): boolean {
+  doesSchemaUtilsDiffer(validator: ValidatorType<T, S, F>, rootSchema: S, behaviorBitFlags: number): boolean {
```
```suggestion
  doesSchemaUtilsDiffer(validator: ValidatorType<T, S, F>, rootSchema: S, defaultFormStateBehavior: number): boolean {
```

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -63,13 +67,18 @@ class SchemaUtils<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends Fo
+    return (
+      this.validator !== validator ||
+      !deepEquals(this.rootSchema, rootSchema) ||
+      this.behaviorBitFlags !== behaviorBitFlags
```
```suggestion
      this. defaultFormStateBehavior !== defaultFormStateBehavior
```

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -87,7 +96,14 @@ class SchemaUtils<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends Fo
+      formData,
+      this.rootSchema,
+      includeUndefinedValues,
+      this.behaviorBitFlags
```
```suggestion
      this. defaultFormStateBehavior
```

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -258,12 +274,17 @@ class SchemaUtils<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends Fo
  *
  * @param validator - an implementation of the `ValidatorType` interface that will be forwarded to all the APIs
  * @param rootSchema - The root schema that will be forwarded to all the APIs
+ * @param [behaviorBitFlags=0] - Optional bitwise flags to set which behavior is chosen for certain edge cases.
```
```suggestion
 * @param [defaultFormStateBehavior=0] - Optional bit flag, if provided, allows users to override default form state behavior
```

heath-freenome commented on packages/utils/src/createSchemaUtils.ts:
```diff
@@ -258,12 +274,17 @@ class SchemaUtils<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends Fo
+>(
+  validator: ValidatorType<T, S, F>,
+  rootSchema: S,
+  behaviorBitFlags = DefaultFormStateBehavior.Legacy_PopulateMinItems
+): SchemaUtilsType<T, S, F> {
+  return new SchemaUtils<T, S, F>(validator, rootSchema, behaviorBitFlags);
```
```suggestion
  defaultFormStateBehavior = DefaultFormStateBehavior.Legacy_PopulateMinItems
): SchemaUtilsType<T, S, F> {
  return new SchemaUtils<T, S, F>(validator, rootSchema, defaultFormStateBehavior);
```

heath-freenome commented on packages/utils/src/schema/getDefaultFormState.ts:
```diff
@@ -92,28 +100,45 @@ function maybeAddDefaultToObject<T = any>(
+  rawFormData?: T;
+  includeUndefinedValues?: boolean | 'excludeObjectChildren';
+  _recurseList?: string[];
+  behaviorBitFlags?: number;
```
```suggestion
  defaultFormStateBehavior?: number;
```

heath-freenome commented on packages/utils/src/schema/getDefaultFormState.ts:
```diff
@@ -306,6 +343,7 @@ export function computeDefaults<T = any, S extends StrictRJSFSchema = RJSFSchema
  * @param [includeUndefinedValues=false] - Optional flag, if true, cause undefined values to be added as defaults.
  *          If "excludeObjectChildren", cause undefined values for this object and pass `includeUndefinedValues` as
  *          false when computing defaults for any nested object properties.
+ * @param [behaviorBitFlags=0] Optional bitwise flags to set which behavior is chosen for certain edge cases.
```
```suggestion
 * @param [defaultFormStateBehavior=0] - Optional bit flag, if provided, allows users to override default form state behavior
```

heath-freenome commented on packages/utils/src/schema/getDefaultFormState.ts:
```diff
@@ -317,14 +355,20 @@ export default function getDefaultFormState<
   rootSchema?: S,
-  includeUndefinedValues: boolean | 'excludeObjectChildren' = false
+  includeUndefinedValues: boolean | 'excludeObjectChildren' = false,
+  behaviorBitFlags = 0
```
```suggestion
  defaultFormStateBehavior = 0
```

heath-freenome commented on packages/utils/src/schema/getDefaultFormState.ts:
```diff
@@ -317,14 +355,20 @@ export default function getDefaultFormState<
+  const defaults = computeDefaults<T, S, F>(validator, schema, {
+    rootSchema,
+    includeUndefinedValues,
+    behaviorBitFlags,
```
```suggestion
    behaviorBitFlags: defaultFormStateBehavior,
```

heath-freenome commented on packages/utils/src/types.ts:
```diff
@@ -956,9 +956,10 @@ export interface SchemaUtilsType<T = any, S extends StrictRJSFSchema = RJSFSchem
    *
    * @param validator - An implementation of the `ValidatorType` interface that will be compared against the current one
    * @param rootSchema - The root schema that will be compared against the current one
+   * @param behaviorBitFlags Bitwise flags to set which behavior is chosen for certain edge cases
```
```suggestion
   * @param [defaultFormStateBehavior=0] - Optional bit flag, if provided, allows users to override default form state behavior
```

heath-freenome commented on packages/utils/src/types.ts:
```diff
@@ -956,9 +956,10 @@ export interface SchemaUtilsType<T = any, S extends StrictRJSFSchema = RJSFSchem
    * @returns - True if the `SchemaUtilsType` differs from the given `validator` or `rootSchema`
    */
-  doesSchemaUtilsDiffer(validator: ValidatorType<T, S, F>, rootSchema: S): boolean;
+  doesSchemaUtilsDiffer(validator: ValidatorType<T, S, F>, rootSchema: S, behaviorBitFlags: number): boolean;
```
```suggestion
  doesSchemaUtilsDiffer(validator: ValidatorType<T, S, F>, rootSchema: S, defaultFormStateBehavior: number): boolean;
```

heath-freenome commented:
@RPiAwesomeness It looks like when you accepted my feedback I wrote it in a way that broke the formatting rules. You can run `npm run cs-format` at the root level to fix it and push a new commit

RPiAwesomeness commented:
I tried running the tests locally after rebasing onto main and correcting the format and it seems like there's a significant number failing all for `antd` and `chakra-ui`. Any ideas as to what might be causing this?

heath-freenome commented:
> I tried running the tests locally after rebasing onto main and correcting the format and it seems like there's a significant number failing all for `antd` and `chakra-ui`. Any ideas as to what might be causing this?

There was a bug fix for antd just pushed over the weekend. Not sure about chakra-ui.

RPiAwesomeness commented:
> There was a bug fix for antd just pushed over the weekend. Not sure about chakra-ui.

Weird. I deleted and re-cloned the repository and now it's passing. Probably some weird caching issue.

heath-freenome commented:
> > There was a bug fix for antd just pushed over the weekend. Not sure about chakra-ui.
> 
> Weird. I deleted and re-cloned the repository and now it's passing. Probably some weird caching issue.

@RPiAwesomeness I recently added a new script to the root to blow away all of the node_modules directories and reinstall them. Next time give that a try?
```
npm run refresh-node-modules
```

nickgros commented:
Remind me why we decided on a bitmask? In my opinion, a bitmask is only useful if you're either
* Using bitwise operators that don't have simple boolean equivalents or
* Are resource constrained, e.g. bandwidth

I don't think that either of these apply here, and I'm worried that some behavior might not decompose to binary options, which would complicate the API even further. I think we would be better off with an object containing options with enum or boolean values. This is basically equivalent to the bitwise but has a more developer-friendly API.

Is the "legacy" behavior going to warrant its own option? Would it make sense for that to just be the `undefined` case for an object prop and make that the default value?

If we're not yet certain about how this new prop API should look, then I think a good option would be to name it something like `experimental_defaultFormStateBehavior` and in the docs state that we may break it between minor/patch versions. That would give us the freedom to wait and see which (if any--there are many other options) of these potential APIs would be best:

```ts
type FormProps = {
  // bitmask
  experimental_defaultFormStateBehavior?: number
}

type FormProps = {
  // enumerated behavior
  experimental_defaultFormStateBehavior?: 'legacy' | 'ignoreArrayMinItemsUnlessRequired'
}

type FormProps = {
  // many individual options
  experimental_defaultFormStateBehavior: {
    arrayMinItems?: 'populate' | 'requiredOnly' // | 'ignore' // if other options like 'ignore' make sense
  }
}

type FormProps = {
  // enums broken down by type
  experimental_defaultFormStateBehavior: {
    arrays?: 'populateMinItems' | 'ignoreMinItemsUnlessRequired'
	objects?: unknown // TBD
  }
}

type FormProps = {
  // boolean fields split by type
  experimental_defaultFormStateBehavior: {
    arrays?: {
      ignoreMinItemsUnlessRequired?: boolean
      // ...other array properties
    }
	objects?: unknown // TBD
  }
}

type FormProps = {
  // enumeration fields split by type
  experimental_defaultFormStateBehavior: {
    arrays?: {
      minItems?: "ignore" | "requiredOnly" | "always"
      // ...other array properties
    }
	objects?: unknown // TBD
  }
}
```

RPiAwesomeness commented:
> Remind me why we decided on a bitmask?
 
@nickgros I believe it was so this could be more flexible in the future because we had initially decided to have a single "features" or "behavior" prop and you could turn individual features on and off something like this: `behavior={DefaultFormStateBehavior.IgnoreMinArrayItemsUnlessRequired | Make.BaconPancakes | Require.AdditionalPylons`. 

I'm personally a bigger fan of switching to a specific type union a la your third example, especially since the configuration of this has already been moved into an individual `defaultFormStateBehavior` prop.

> ```ts
> type FormProps = {
>   // many individual options
>   experimental_defaultFormStateBehavior: {
>     arrayMinItems?: 'populate' | 'requiredOnly' // | 'ignore' // if other options like 'ignore' make sense
>   }
> }
> ```

I remember something about adding at least one more mode but I don't remember what exactly at this moment. I'll wait for input from @heath-freenome before finally making the changes but I think that implementation makes a lot more sense in terms of developer ergonomics.

heath-freenome commented:
> > Remind me why we decided on a bitmask?
> 
> @nickgros I believe it was so this could be more flexible in the future because we had initially decided to have a single "features" or "behavior" prop and you could turn individual features on and off something like this: `behavior={DefaultFormStateBehavior.IgnoreMinArrayItemsUnlessRequired | Make.BaconPancakes | Require.AdditionalPylons`.
> 
> I'm personally a bigger fan of switching to a specific type union a la your third example, especially since the configuration of this has already been moved into an individual `defaultFormStateBehavior` prop.
> 
> > ```ts
> > type FormProps = {
> >   // many individual options
> >   experimental_defaultFormStateBehavior: {
> >     arrayMinItems?: 'populate' | 'requiredOnly' // | 'ignore' // if other options like 'ignore' make sense
> >   }
> > }
> > ```
> 
> I remember something about adding at least one more mode but I don't remember what exactly at this moment. I'll wait for input from @heath-freenome before finally making the changes but I think that implementation makes a lot more sense in terms of developer ergonomics.

Let's go with the `experimental_defaultFormStateBehavior` solution

RPiAwesomeness commented:
I've made the feature flag refactor and rebased onto main, should be working the same as expected. Minor changes to the playground to support the new feature flag format and updated the documentation.

heath-freenome commented:
@RPiAwesomeness it looks like you'll need to run `npm run cs-format` for your PR

heath-freenome commented on CHANGELOG.md:
```diff
@@ -15,6 +15,19 @@ it according to semantic versioning. For example, if your PR adds a breaking cha
 should change the heading of the (upcoming) version to include a major version bump.
 
 -->
+# 5.8.0
```
Since we haven't published `5.7.0` yet this could be released as part of that release. I have one more PR to complete for it and won't be done with it until early next week.

RPiAwesomeness commented:
Re-ran `cs-format` and updated the CHANGELOG

nickgros commented on packages/utils/src/schema/getDefaultFormState.ts:
```diff
@@ -93,28 +100,45 @@ function maybeAddDefaultToObject<T = any>(
   }
 }
 
+interface computeDefaultsProps<T = any, S extends StrictRJSFSchema = RJSFSchema> {
```
For consistency, all interfaces and types use PascalCase.

```suggestion
interface ComputeDefaultsProps<T = any, S extends StrictRJSFSchema = RJSFSchema> {
```

nickgros commented on packages/utils/src/types.ts:
```diff
@@ -31,6 +31,10 @@ export type RJSFSchema = StrictRJSFSchema & GenericObjectType;
  */
 export type FormContextType = GenericObjectType;
 
+export type Experimental_DefaultFormStateBehavior = {
+  arrayMinItems?: 'populate' | 'requiredOnly';
+};
```
Would like this types documentation to match what's in the docs project, not a big deal while this is experimental though

nickgros replied:
I just mean having the supplemental text here as well so IDEs can provide the documentation info at a glance, so one does not have to open the docs page to know what each option does

nickgros commented on packages/utils/test/schema/getDefaultFormStateTest.ts:
```diff
@@ -34,7 +34,7 @@ export default function getDefaultFormStateTest(testValidator: TestValidatorType
           $ref: '#/definitions/testdef',
         };
-        expect(computeDefaults(testValidator, schema, undefined, schema)).toEqual({
+        expect(computeDefaults(testValidator, schema, { rootSchema: schema })).toEqual({
```
I wasn't sold the new function signature for `computeDefaults` until I saw these tests. This is much cleaner!

RPiAwesomeness replied:
With all the new optional stuff being added between my MR and the others going in ahead of me I figured it just made sense. Glad you like it!

nickgros commented on packages/utils/test/schema/getDefaultFormStateTest.ts:
```diff
@@ -218,20 +228,124 @@ export default function getDefaultFormStateTest(testValidator: TestValidatorType
+        ).toEqual({ requiredArray: ['default0', 'default1'] });
+      });
+      // BUG: https://github.com/rjsf-team/react-jsonschema-form/issues/3602
+      it.skip('should combine defaults with raw form data for a required array property with minItems', () => {
```
Thanks for keeping the skipped test so this is easy to come back to

nickgros commented on packages/utils/test/schema/getDefaultFormStateTest.ts:
```diff
@@ -218,20 +228,124 @@ export default function getDefaultFormStateTest(testValidator: TestValidatorType
+          })
+        ).toEqual({});
+      });
+      it('should return empty array when given an empty array as form data for an optional array property with minItems ', () => {
```
```suggestion
      it('should return empty array when given an empty array as form data for an optional array property with minItems', () => {
```

nickgros commented on packages/docs/docs/api-reference/form-props.md:
```diff
@@ -59,6 +59,41 @@ Formerly the `validate` prop.
+  <Form
+    schema={schema}
+    validator={validator}
+    experimental_defaultFormStateBehavior={
+      arrayMinItems: 'requiredOnly'
+    }
```
```suggestion
    experimental_defaultFormStateBehavior={{
      arrayMinItems: 'requiredOnly'
    }}
```

nickgros reviewed (approved):
Couple of suggestions but none of them should block us from merging this feature. This looks great!
